### PR TITLE
8332474: Tighten up ToolBox' JavacTask to not silently accept javac crash as a failure

### PR DESCRIPTION
**Pull request by lahodaj**
Tests for javac use several test frameworks, and one of them is the "toolbox", which provides `JavacTask` that allows to conveniently run javac on a given set of sources. The problem with `JavacTask` is that, by default, when a compilation failure is expected, javac crash (exit code 4) is tolerated by the `JavacTask`. And the test must manually select a specific exit code to overwrite this behavior.

This then leads to bugs like https://bugs.openjdk.org/browse/JDK-8335385, which are silently ignored by `JavacTask`.

The proposal herein is to tighten up the `JavacTask`, and effectively disallow exit code 4 for `JavacTask` (but permit any other exit code, as javac is using several exit codes). The base implementation in `AbstractTask` is changed to use a validator for the exit codes, which is then leveraged by `JavacTask`.

This patch depends on PR #19969, as that fixes JDK-8335385, where the javac crash is ignored. It also tweaks module attribution to not leave empty `JCModuleDecl.sym` for duplicate modules, and set it to an erroneous module. The empty (`null`) symbol here crashes javac in the `test/langtools/tools/javac/modules/MultiModuleModeTest.java#testDuplicateModules`, and the test wouldn't pass with this more strict `JavacTask`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332474](https://bugs.openjdk.org/browse/JDK-8332474): Tighten up ToolBox' JavacTask to not silently accept javac crash as a failure (**Task** - P4)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19972/head:pull/19972` \
`$ git checkout pull/19972`

Update a local copy of the PR: \
`$ git checkout pull/19972` \
`$ git pull https://git.openjdk.org/jdk.git pull/19972/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19972`

View PR using the GUI difftool: \
`$ git pr show -t 19972`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19972.diff">https://git.openjdk.org/jdk/pull/19972.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19972#issuecomment-2200267662)

### CODE REVIEW

File: test/langtools/tools/lib/toolbox/AbstractTask.java [hunk 1]
```diff
@@ -83,17 +86,56 @@ public Result run(Expect expect, int exitCode) {
         return run();
     }
 
+    /**
+     * Sets the expected outcome of the task and calls {@code run()}.
+     * @param expect the expected outcome
+     * @param exitCodeValidator an exit code validator. The first parameter will
+     *                          be the actual exit code, the second test name,
+     *                          should throw TaskError is the exit code is not
```
Comment (jonathan-gibbons) [inline test/langtools/tools/lib/toolbox/AbstractTask.java:94]: typo: "is the exit code" -> "if the exit code"

File: test/langtools/tools/lib/toolbox/JavacTask.java [hunk 2]
```diff
@@ -314,6 +314,26 @@ public String name() {
...
+            if (exitCode == 4) {
+                throw new TaskError("Task " + testName + " failed due to a javac crash "
+                    + "(exit code 4)");
+            }
+        });
+    }
+
+    @Override
+    public Result run(Expect expect, int exitCode) {
+        if (exitCode == 4) {
```
Comment (jonathan-gibbons) [inline test/langtools/tools/lib/toolbox/JavacTask.java:331]: I guess we never expect `javac` to crash, and should not be able to test that it does, right?

File: test/langtools/tools/lib/toolbox/AbstractTask.java (hunk 1 above)
Comment (lahodaj) [reply test/langtools/tools/lib/toolbox/AbstractTask.java]: Thanks, fixed now!

File: test/langtools/tools/lib/toolbox/JavacTask.java (hunk 2 above)
Comment (lahodaj) [reply test/langtools/tools/lib/toolbox/JavacTask.java:331]: I am not too worried about a test that would intentionally crash javac, and would need to test for that. It is not clear to me what would be the purpose of the test. And, assuming this need would indeed be rare, there are other ways to test for that. So, I opted for a more strict approach. OTOH, I don't have a too strong opinion on this, so I can re-allow exit code 4 here if desired.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Pull request closed without merging**